### PR TITLE
`sage.combinat.root_system`: Modularization fixes, `# needs`

### PR DESCRIPTION
**Pull request by mkoeppe**
Most of the package `sage.combinat.root_system` is intended to be shipped by the distribution **sagemath-modules**, and much of the functionality can be tested in the modularized setting. 

- Some functionality depends on other distributions, in particular **sagemath-graphs** (via Dynkin diagrams). The doctests are marked as `# needs sage.graphs`.
- Other doctests that depend on combinatorics functionality such as Tableaux are marked `# needs sage.combinat`.
- Use of Weyl and Coxeter groups is marked `# needs sage.libs.gap`, `# needs sage.rings.number_field`, `# needs sage.symbolic`, or `# needs sage.groups`.
- Some imports are changed to `lazy_import`; in some places, the changed code uses the fact that `isinstance(x, cls)` gives False for unresolved lazy import `cls`.
- Some `_test_...` and `some_elements` methods are changed so that `ImportError`s do not lead to test failures.
- The hardcoded dependence of `CartanMatrix` on `Matrix_integer_sparse` (which has complicated compile-time dependencies) is relaxed.
- doctest cosmetics: Breaking long lines in doctest input and output so that it can be read in the HTML documentation without (or with less) horizontal scrolling

- Part of: #29705 
- Cherry-picked from: #35095 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/combinat/root_system/cartan_type.py [hunk 1]
```diff
@@ -335,27 +333,27 @@
 
     sage: CartanType(["A", 9, 2])
     ['B', 5, 1]^*
-    sage: CartanType(["A", 9, 2]).dynkin_diagram()
+    sage: CartanType(["A", 9, 2]).dynkin_diagram()                                      # needs sage.graphs
```
Comment (dcoudert) [inline src/sage/combinat/root_system/cartan_type.py]: could use block scope here

File: src/sage/combinat/root_system/cartan_type.py [hunk 2]
```diff
@@ -365,27 +363,27 @@
     sage: CartanType.options['notation'] = 'Kac'
     sage: CartanType(["A", 9, 2])
     ['A', 9, 2]
-    sage: CartanType(["A", 9, 2]).dynkin_diagram()
+    sage: CartanType(["A", 9, 2]).dynkin_diagram()                                      # needs sage.graphs
```
Comment (dcoudert) [inline src/sage/combinat/root_system/cartan_type.py]: block scope

File: src/sage/combinat/root_system/extended_affine_weyl_group.py [hunk 3]
```diff
@@ -1370,8 +1418,8 @@ def apply_simple_projection(self, i, side='right', length_increasing=True):
 
                 - ``self`` -- an element of the extended affine Weyl group
                 - `i` -- a Dynkin node (index of a simple reflection `s_i`)
-                - ``side`` -- 'right' or 'left' (default: 'right') according to which side of ``self`` the reflection `s_i` should be multiplied
-                - ``length_increasing`` -- True or False (default True). If False do the above with the word "greater" replaced by "less".
+                - ``side`` -- ``'right'`` or ``'left'`` (default: ``'right'``) according to which side of ``self`` the reflection `s_i` should be multiplied
```
Comment (dcoudert) [inline src/sage/combinat/root_system/extended_affine_weyl_group.py]: 80 columns mode ?

File: src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py [hunk 4]
```diff
@@ -421,99 +433,108 @@ class NonSymmetricMacdonaldPolynomials(CherednikOperatorsEigenvectors):
         sage: KL0 = KL.classical()
         sage: P = K['x0,x1']
         sage: def EE(weight): return E[L0(weight)].expand(P.gens())
+
+        sage: # needs sage.combinat
         sage: import sage.combinat.sf.ns_macdonald as NS
         sage: EE([0,0])
         1
-        sage: NS.E([0,0])
+        sage: NS.E([0,0])                                                               # needs sage.groups
```
Comment (dcoudert) [inline src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py:441]: could use block scope for `sage.groups` too

File: src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py [hunk 5]
```diff
@@ -791,7 +857,7 @@ class NonSymmetricMacdonaldPolynomials(CherednikOperatorsEigenvectors):
...
...^2*q2-q^7*q1*q2^2-q^7*q2^3-2*q^5*q1^2*q2-4*q^5*q1*q2^2-2*q^5*q2^3-2*q^3*q1^2*q2-4*q^3*q1*q2^2-2*q^3*q2^3)/(q^8*q1^3+q^6*q1^2*q2-q^2*q1*q2^2-q2^3))*B[(1, 0)] + ((q^6*q1^2*q2+2*q^6*q1*q2^2+q^6*q2^3+q^4*q1^2*q2+2*q^4*q1*q2^2+q^4*q2^3)/(-q^8*q1^3-q^6*q1^2*q2+q^2*q1*q2^2+q2^3))*B[(1, -1)] + ((q^8*q1^3+q^8*q1^2*q2+q^6*q1^3+q^6*q1^2*q2-q^6*q1*q2^2-q^6*q2^3-2*q^4*q1^2*q2-4*q^4*q1*q2^2-2*q^4*q2^3-q^2*q1^2*q2-3*q^2*q1*q2^2-2*q^2*q2^3)/(q^8*q1^3+q^6*q1^2*q2-q^2*q1*q2^2-q2^3))*B[(1, 1)] + ((q^5*q1^2+q^5*q1*q2-q^3*q1*q2-q^3*q2^2-q*q1*q2-q*q2^2)/(q^6*q1^2-q2^2))*B[(1, 2)] + ((-q^6*q1^2-q^6*q1*q2+q^4*q1*q2+q^4*q2^2+q^2*q1*q2+q^2*q2^2)/(-q^6*q1^2+q2^2))*B[(2, 0)] + ((-q^3*q1*q2-q^3*q2^2)/(q^6*q1^2-q2^2))*B[(2, -1)] + ((-q^5*q1^2-q^5*q1*q2+q^3*q1*q2+q^3*q2^2+q*q1*q2+q*q2^2)/(-q^6*q1^2+q2^2))*B[(2, 1)] + B[(2, 2)] + ((q^7*q1^2*q2+2*q^7*q1*q2^2+q^7*q2^3+q^5*q1^2*q2+2*q^5*q1*q2^2+q^5*q2^3)/(-q^8*q1^3-q^6*q1^2*q2+q^2*q1*q2^2+q2^3))*B[(0, -1)] + ((q^7*q1^3+q^7*q1^2*q2-q^7*q1*q2^2-q^7*q2^3-2*q^5*q1^2*q2-4*q^5*q1*q2^2-2*q^5*q2^3-2*q^3*q1^2*q2-4*q^3*q1*q2^2-2*q^3*q2^3)/(q^8*q1^3+q^6*q1^2*q2-q^2*q1*q2^2-q2^3))*B[(0, 1)] + ((q^6*q1^2+q^6*q1*q2-q^4*q1*q2-q^4*q2^2-q^2*q1*q2-q^2*q2^2)/(q^6*q1^2-q2^2))*B[(0, 2)]
```
Comment (dcoudert) [inline src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py]: may be this output deserves a better alignment (quite long).

File: src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py [hunk 6]
```diff
@@ -1074,29 +1162,29 @@ class NonSymmetricMacdonaldPolynomials(CherednikOperatorsEigenvectors):
...
...t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x3)+(t+1)*(t-1)*x2^2*x1/((t*q-1)*(t*q+1)*x3)+(t-1)^3*(t+1)^3*(1+t^2+t^4*q^2)*q*x3*x1^2/((t^3*q^2-1)*(t^3*q^2+1)*(t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1))+(t-1)*(t+1)*q^3/((q^3*t^5+1)*(q^3*t^5-1)*x1*x2*x3)+(t-1)^2*(t+1)^2*(3+3*q^2+q^4+2*q^2*t^2-t^2+t^2*q^4-6*t^4*q^2+q^4*t^4-8*t^6*q^4+q^2*t^6+2*t^8*q^4-4*t^8*q^6+t^10*q^4+2*t^10*q^6-2*q^8*t^12-t^14*q^10+t^14*q^8+q^8*t^16+q^10*t^16+2*q^10*t^18)*q^2/((q^3*t^5+1)*(q^3*t^5-1)*(t*q-1)*(t*q+1)*(t^3*q^2+1)*(t^3*q^2-1)*(t^2*q-1)*(t^2*q+1))+(t-1)^2*(t+1)^2*(-q^4-2*q^2-1-t^2*q^4-t^4*q^6+2*q^6*t^6+t^6*q^4+t^10*q^6+q^8*t^12+t^14*q^10)*q/((t^3*q^2-1)*(t^3*q^2+1)*(t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x3)+(t-1)^2*(t+1)^2*(-1-q^2-q^2*t^2+t^2+t^4*q^2-q^4*t^4+2*t^6*q^4)*q*x3*x1/((t^3*q^2-1)*(t^3*q^2+1)*(t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x2)+(t-1)^2*(t+1)^2*x2*x1^2/((t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x3)+(t-1)^2*(t+1)^2*x3*x1^2/((t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x2)+(t-1)^2*(t+1)^2*q^4/((t*q+1)*(t*q-1)*(q^3*t^5+1)*(q^3*t^5-1)*x1*x2)+(t-1)^2*(t+1)^2*(-q^2-1-q^2*t^2-q^4*t^4+t^6*q^4+t^10*q^6+q^8*t^12+t^14*q^10)*q*x3*x2/((t^3*q^2-1)*(t^3*q^2+1)*(t*q+1)*(t*q-1)*(q^3*t^5-1)*(q^3*t^5+1)*x1)  # needs sage.symbolic
```
Comment (dcoudert) [inline src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py:1196]: Waouh, this one is long.

File: src/sage/combinat/root_system/root_space.py [hunk 7]
```diff
@@ -328,7 +328,7 @@ def associated_coroot(self):
 
     def quantum_root(self):
         r"""
-        Returns True if ``self`` is a quantum root and False otherwise.
+        Return ``True`` if ``self`` is a quantum root and ``False`` otherwise.
```
Comment (dcoudert) [inline src/sage/combinat/root_system/root_space.py]: Why not changing to a formulation like `Check whether ``self`` is a quantum root`

File: src/sage/combinat/root_system/root_space.py [hunk 8]
```diff
@@ -411,7 +412,7 @@ def max_coroot_le(self):
 
     def max_quantum_element(self):
         r"""
-        Returns a reduced word for the longest element of the Weyl group whose shortest path in the quantum Bruhat graph to the identity, has sum of quantum coroots at most ``self``.
+        Return a reduced word for the longest element of the Weyl group whose shortest path in the quantum Bruhat graph to the identity, has sum of quantum coroots at most ``self``.
```
Comment (dcoudert) [inline src/sage/combinat/root_system/root_space.py]: 80 cols mode

File: src/sage/combinat/root_system/type_B_affine.py [hunk 9]
```diff
@@ -54,29 +54,28 @@ def dynkin_diagram(self):
 
         EXAMPLES::
 
-            sage: b = CartanType(['B',3,1]).dynkin_diagram()
-            sage: b
+            sage: b = CartanType(['B',3,1]).dynkin_diagram(); b                         # needs sage.graphs
```
Comment (dcoudert) [inline src/sage/combinat/root_system/type_B_affine.py]: block scope

File: src/sage/combinat/root_system/weight_lattice_realizations.py [hunk 10]
```diff
@@ -648,14 +667,17 @@ def _test_reduced_word_of_translation(self, elements=None, **options):
...
 
             See the documentation for :class:`TestSuite` for more information.
             """
             tester = self._tester(**options)
             if not self.cartan_type().is_affine(): # won't be necessary anymore once root systems are categorified
                 return
-            alpha = self.simple_roots()
+            try:
+                alpha = self.simple_roots()
+            except ImportError:  # when sagemath-graphs is not available
```
Comment (dcoudert) [inline src/sage/combinat/root_system/weight_lattice_realizations.py]: `sagemath-graphs` or `sage.graphs` ?

**Review by dcoudert — CHANGES_REQUESTED** [review]
a few remarks

File: src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py (hunk 4 above)
Comment (mkoeppe) [reply src/sage/combinat/root_system/non_symmetric_macdonald_polynomials.py:441]: Here doctests that need sage.groups are alternating with doctests that don't; this information would be lost with a block tag

**Review by dcoudert — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by vbraun** [comment]
pdf docs fail

**Comment by mkoeppe** [comment]
Sorry, fixed now

**Pull request merged**